### ISSUE TITLE
Usage of resumable exceptions?

### ISSUE THREAD

**Issue by suynwa** [SAP/styleguides#189]
The OP has highlighted in this blog https://blogs.sap.com/2021/02/02/write-smart-abap-not-boring-abap-part-1/ a valid use-case for resumable exceptions.

The styleguides don't mention about resumable exceptions. What are your thoughts about them?

**Comment by jordao76** [comment]
Not a fan. I see resumable exceptions as an exotic feature with no applicability in modern business application design, which is the main focus of ABAP. Exceptions should not be used to control the normal flow of a program, and that's what resumable exceptions allow you to do. Their handling gets intertwined in the normal flow of your program.

**Comment by pokrakam** [comment]
They have their use, but IMHO should be used very, very sparingly. They are useful if the caller needs to be able to decide if an exception is OK. Usually there are better solutions, for example I believe the example in the blog could be achieved more elegantly with an interface + factory + different implementing classes. 

One use case is a lock error on an expensive read: We leave it up to the caller if they want to fail outright and skip the read, or continue without lock - e.g. switch to a readonly scenario - and RESUME to go ahead and read the locked data.

Their big drawback is that the code easily becomes tangled and hard to follow. Consider the execution sequence, numbered here for your convenience: 
```
METHOD caller.
  TRY. 
      foo( ).              "1
      next( ).             "5
    CATCH.
      handle_error( ).     "3
      RESUME. 
  ENDTRY.
ENDMETHOD. 

METHOD foo.
  do_thing_that_fails( ).  "2
  do_more_stuff( ).        "4
ENDMETHOD.
```
Now picture the same but with more code in each block. It can get messy. 
On my travels I once encountered a construct where the CATCH block itself raised another RESUMABLE exception, so we had the above sequence expanded across three layers of calls. I had to diagram it to understand what was going on, shared here from my notes: 
![image](https://user-images.githubusercontent.com/25539927/113516590-f7bb7500-9572-11eb-98ca-8942fc04e48a.png)

**Comment by larshp** [comment]
personally, I think code should be boring and simple over smart and exciting

boring code is easier to maintain and extend, and sometimes it takes more effort to write simple code than finding an exciting feature to use

**Comment by suynwa** [comment]
> personally, I think code should be boring and simple
Says the man who single handedly brought open Source projects to the SAP developer community

**Comment by HrFlorianHoffmann** [comment]
The only case where my team ever discussed applying resumable exceptions was a series of validations. We wanted to validate user input, and report back all bad input at once instead of stopping after the first mistake and serving our feedback back to the user slice by slice in "salami tactics".

We considered using the following pattern:

    METHOD validate_user_input.
      TRY.
          validate_id( user_input ).
          validate_name( user_input ).
          validate...( ... ).
        CATCH bad_input INTO lx_bad_input.
          INSERT lx_bad_input->get_text( ) INTO TABLE mistakes.
          RESUME.
      ENDTRY.
    ENDMETHOD.

    METHOD validate_id.
      IF user_input-id IS INITIAL.
        RAISE NEW bad_input( 'No ID provided' ).
      ENDIF.
    ENDMETHOD.

In the end we decided against it. The pattern works but we found using exceptions for it too heavy: For a method that has "validate" in its name, the result "invalid" is a regular outcome. Something that should be handled by a regular return value like "false". Definitely not something "unexpected" that should be represented by exceptions.

So in the end we decided for more regular patterns:

    METHOD validate_user_input.
        INSERT LINES OF validate_id( user_input ) INTO TABLE mistakes.
        INSERT LINES OF validate_name( user_input ) INTO TABLE mistakes.
        INSERT LINES OF validate_...( ... ) INTO TABLE mistakes.
    ENDMETHOD.

    METHOD validate_id.
      IF user_input-id IS INITIAL.
        INSERT `No ID provided` INTO TABLE result.
      ENDIF.
    ENDMETHOD.

**Comment by pokrakam** [comment]
@HrFlorianHoffmann a perfect example of 'Usually there are better solutions'. The second version is both shorter and simpler to follow. Also in that example, there was no choice on the exception handler's part whether to resume or not.

**Comment by suynwa** [comment]
Closing the issue with the following TLDR: 
Using resumable exception should not be a _norm_, since they allow the exception handler to influence the program flow. They should be used in exceptional situations only.
